### ISSUE TITLE
Build from source: CMake not using compilers installed by Anaconda, fails with stl_pair.h errors

### ISSUE THREAD

**Issue by wpm** [pytorch/pytorch#47717]
## 🐛 Bug

I tried to build PyTorch from source in an Anaconda environment, using Anaconda to update to the latest C and C++ compilers. PyTorch did not build because it ignored the upgraded compilers. When I forced it to use the upgraded compilers, it failed when building the Caffee components with errors about `stl_pair.h`.

## To Reproduce

1. `conda create --name pytorch -y`
2. `conda activate pytorch`
3. `conda install -c anaconda gcc_linux-64 gxx_linux-64 -y`
4. `conda install numpy ninja pyyaml mkl mkl-include setuptools cmake cffi typing_extensions future six requests dataclasses -y`
5. `conda install -c pytorch magma-cuda110 -y`
6. `git clone --recursive https://github.com/pytorch/pytorch`
7. `cd pytorch`
6. `git submodule sync`
7. `git submodule update --init --recursive`
8. `export CMAKE_PREFIX_PATH=${CONDA_PREFIX:-"$(dirname $(which conda))/../"}`
9. `python setup.py install`

Everything except step (3) is taken from the PyTorch instructions for [installing from source](https://github.com/pytorch/pytorch#from-source). Step (3) uses Anaconda to [upgrade the C and C++ compilers](https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html).

As described in the Anaconda compiler tools documentation, step (3) installs compilers under names decorated with host and target platform names and updates the `$CC`, `$CXX`, `$GCC`, and `$GXX` environment variables accordingly. On my machine these are as follows.

```
(pytorch) $ echo $CC
/home/wmcneill/anaconda3/envs/pytorch/bin/x86_64-conda_cos6-linux-gnu-cc
(pytorch) $ echo $CXX
/home/wmcneill/anaconda3/envs/pytorch/bin/x86_64-conda_cos6-linux-gnu-c++
(pytorch) $ echo $GCC
/home/wmcneill/anaconda3/envs/pytorch/bin/x86_64-conda_cos6-linux-gnu-gcc
(pytorch) $ echo $GXX
/home/wmcneill/anaconda3/envs/pytorch/bin/x86_64-conda_cos6-linux-gnu-g++
```

These are all version 7.3.0.

The build fails immediately when checking compiler versions. The PyTorch compiler check runs on the Anaconda default `/home/wmcneill/anaconda3/bin/c++` instead of the newer compilers installed by Anaconda. That compiler is version 5.2.0.

```
-- Check for working CXX compiler: /home/wmcneill/anaconda3/bin/c++
-- Check for working CXX compiler: /home/wmcneill/anaconda3/bin/c++ - broken
CMake Error at /home/wmcneill/anaconda3/envs/pytorch/share/cmake-3.18/Modules/CMakeTestCXXCompiler.cmake:59 (message):
  The C++ compiler

    "/home/wmcneill/anaconda3/bin/c++"

  is not able to compile a simple test program.
```

The full output is here: [pytorch-compilation-error.txt](https://github.com/pytorch/pytorch/files/5520836/pytorch-compilation-error.txt).

If I manually change the compilers in `/home/wmcneill/anaconda3/bin` to point to the newly upgraded ones I get past the compiler checks, but get 91 error messages when compiling the Caffe component. They appear to occur for most of the `*.cu` source files in `pytorch/aten/src/ATen/native/cuda`. A typical error looks like this.

```
[4517/5577] Building NVCC (Device) object caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/torch_cuda_generated_THCTensorIndex.cu.o
FAILED: caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/torch_cuda_generated_THCTensorIndex.cu.o 
cd /home/wmcneill/src/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC && /home/wmcneill/anaconda3/envs/pytorch/bin/cmake -E make_directory /home/wmcneill/src/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/. && /home/wmcneill/anaconda3/envs/pytorch/bin/cmake -D verbose:BOOL=OFF -D build_configuration:STRING=Release -D generated_file:STRING=/home/wmcneill/src/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/./torch_cuda_generated_THCTensorIndex.cu.o -D generated_cubin_file:STRING=/home/wmcneill/src/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/./torch_cuda_generated_THCTensorIndex.cu.o.cubin.txt -P /home/wmcneill/src/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/torch_cuda_generated_THCTensorIndex.cu.o.Release.cmake
/home/wmcneill/anaconda3/envs/pytorch/x86_64-conda_cos6-linux-gnu/include/c++/7.3.0/bits/stl_pair.h(437): error: argument list for class template "std::pair" is missing

/home/wmcneill/anaconda3/envs/pytorch/x86_64-conda_cos6-linux-gnu/include/c++/7.3.0/bits/stl_pair.h(437): error: expected a ")"

/home/wmcneill/anaconda3/envs/pytorch/x86_64-conda_cos6-linux-gnu/include/c++/7.3.0/bits/stl_pair.h(437): error: template parameter "_T1" may not be redeclared in this scope

/home/wmcneill/anaconda3/envs/pytorch/x86_64-conda_cos6-linux-gnu/include/c++/7.3.0/bits/stl_pair.h(437): error: expected a ";"
[...]
```

## Expected behavior

I expect the PyTorch build system to use the `$CC`, `$CXX`, `$GCC`, and `$GXX` environment variables which point to the newer compiler versions.

When I manually change the compilers to the new versions I expect PyTorch to build and install.

## Environment

```
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
Clang version: Could not collect
CMake version: version 3.18.4

Python version: 3.9 (64-bit runtime)
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla P100-PCIE-16GB
GPU 1: Tesla P100-PCIE-16GB

Nvidia driver version: 455.32.00
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[conda] magma-cuda110             2.5.2                         1    pytorch
[conda] mkl                       2020.4             h726a3e6_304    conda-forge
[conda] mkl-include               2020.4             h726a3e6_304    conda-forge
[conda] numpy                     1.19.4           py39h57d35e7_1    conda-forge
```

The GCC version comes from the default system compiler installed at `/usr/bin/gcc`.

## Additional context

I first encountered build problems in #47529. That involved git errors which have since been resolved. Following a request from @ptrblck, I am opening this new issue to focus exclusively on the compilation problems.

cc @malfet @seemethere @walterddr

**Comment by skyline75489** [comment]
According to this [documentation](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#axzz3x7mwvZrG) from NVIDIA, the supported GCC version in all Linux distribution is GCC 9. 

It mentions that:

>Note that starting with CUDA 11.0, the minimum recommended GCC compiler is at least GCC 5 due to C++11 requirements in CUDA libraries e.g. cuFFT and CUB

But in the support matrix there's only GCC 9. I guess using GCC 7.3 might be the root cause of the problem.

And NVIDIA did not list Anaconda GCC toolchain distribution as supported. So I wouldn't really count on it. 

---

Based on your environment (CentOS 7 with GCC 4.8.5) I'm guessing you are having a hard time compiling PyTorch with your native Linux environment. So you want to give Anaconda a try. Sadly it seems that GCC 7.3 is also not good enough.

**Comment by wpm** [comment]
There was a problem with the compiler tool chain. After fixing that I get a different error when trying to link to a CUDA library.

### Compiler Tool Chain Fix

It turns out version 9.3.0 compilers are installed in the `anaconda3/bin` directory but when I activate an environment other than base the environment variables pointing to them are removed. I'm not sure if that is by design or a problem with my conda  installation, but it can be fixed by simply setting

```
export CC=/home/wmcneill/anaconda3/bin/x86_64-conda-linux-gnu-cc
export CXX=/home/wmcneill/anaconda3/bin/x86_64-conda-linux-gnu-c++
export GCC=/home/wmcneill/anaconda3/bin/x86_64-conda-linux-gnu-gcc
export GXX=/home/wmcneill/anaconda3/bin/x86_64-conda-linux-gnu-g++
```

It is also not necessary to install the gcc_linux-64 and gxx_linux-64 compiler packages that I list in step (3) of the repro steps.

### libcusparse.so.11 Linker Error

When I run `python setup.py install` it gets further. The correct compilers are used.

```
Building wheel torch-1.8.0a0
-- Building version 1.8.0a0
cmake -GNinja -DBUILD_PYTHON=True -DBUILD_TEST=True -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/home/wmcneill/src/pytorch/torch -DCMAKE_PREFIX_PATH=/home/wmcneill/anaconda3/envs/layoutlm -DNUMPY_INCLUDE_DIR=/home/wmcneill/anaconda3/envs/layoutlm/lib/python3.7/site-packages/numpy/core/include -DPYTHON_EXECUTABLE=/home/wmcneill/anaconda3/envs/layoutlm/bin/python -DPYTHON_INCLUDE_DIR=/home/wmcneill/anaconda3/envs/layoutlm/include/python3.7m -DPYTHON_LIBRARY=/home/wmcneill/anaconda3/envs/layoutlm/lib/libpython3.7m.so.1.0 -DTORCH_BUILD_VERSION=1.8.0a0 -DUSE_NUMPY=True /home/wmcneill/src/pytorch
-- The CXX compiler identification is GNU 9.3.0
-- The C compiler identification is GNU 9.3.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /home/wmcneill/anaconda3/bin/x86_64-conda-linux-gnu-c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/wmcneill/anaconda3/bin/x86_64-conda-linux-gnu-cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
[...]
```

However, there is a different error when building Caffe. I get several warnings like this

```
/home/wmcneill/anaconda3/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: warning: libcusparse.so.11, needed by /home/wmcneill/src/pytorch/build/lib/libtorch_cuda.so, not found (try using -rpath or -rpath-link)
```

Eventually the link step fails. Full error log [here](https://github.com/pytorch/pytorch/files/5526807/pytorch-compile-log.txt.gz).

The `libcusparse.so.11` library is present in `$CUDA_HOME/targets/x86_64-linux/lib/libcusparse.so.11`. Do I need to set the `RPATH`? How do I do that?

**Comment by pearu** [comment]
Note that
```
conda install -c anaconda gcc_linux-64 gxx_linux-64 -y
```
gives gcc 7. I suggest using
```
conda install -c conda-forge gcc_linux-64 gxx_linux-64 -y
```
that will install gcc 9. In fact, the subsequent `conda install ..` will upgrade gcc 7 to gcc 9 anyway.

Pytorch requires C++14, so one must apply:
```
export CXXFLAGS="`echo $CXXFLAGS | sed 's/-std=c++17/-std=c++14/'`"
```

[Perhaps OT] Locally, when trying to reproduce this issue, I noticed that `python` refers to python in conda base environment rather than the python in the given pytorch conda environment. This leads to other issues (cannot import `typing_extensions`) when running `python setup.py develop`. So, make sure that the correct python is picked up.

Next, I got the following build failure [perhaps also OT]:
```
../third_party/tensorpipe/tensorpipe/common/shm_segment.cc: In function 'std::tuple<tensorpipe::Error, tensorpipe::Fd> tensorpipe::{anonymous}::openTmpfileInDevShm()':
../third_party/tensorpipe/tensorpipe/common/shm_segment.cc:79:15: error: 'O_TMPFILE' was not declared in this scope
   79 |   int flags = O_TMPFILE | O_EXCL | O_RDWR | O_CLOEXEC;
      |               ^~~~~~~~~
```
To overcome this as an unrelated build issue, I set `USE_TENSORPIPE=0` and restarted the build from a clean state (`git clean -xddf; git submodule foreach --recursive git clean -xfdd`).

Now the build seems to be fine.

In summary, when using conda environment, make sure that C++14 is effective:
```
export CXXFLAGS="`echo $CXXFLAGS | sed 's/-std=c++17/-std=c++14/'`"
```
```

**Comment by walterddr** [comment]
Not just that 
1. since pytorch uses cmake_cxx_standard, you should actually remove `-std=c++xx` all together in your env CXXFLAGS because cmake will do the configuration for you.
2. you need to try multiple times and see where are all those linker error coming from. Unfortunately conda toolchain doesn't package everything and sometimes you still need to refer to your native system toolchain for libraries and binaries. 
  - you probably need `-Wl,-rpath,$CUDA_HOME/lib64` and `-Wl,-rpath-link,$CUDA_HOME/lib64` 
  - your native library path `-L/lib` or `-L/lib64`
  - you also probably need to alter your you LD_LIBRARY_PATH accordingly as well.
3. for the conda env: explicitly setting cmake python settings is useful here such as `PYTHON_EXECUTABLE`

Regarding the Tensorpipe issue, it is reported in #56777

**Comment by pearu** [comment]
Thanks, @walterddr , for pointing out other issues and solutions. Indeed, specifying CUDA_HOME is needed to resolve the linker error (I have updated my comment accordingly).

**Comment by walterddr** [comment]
glad that it worked. I am closing this issue as there's a "workaround". Please feel free to reopen it if the instruction above doesn't resolve your situation.
